test/update_cr/update_cr_test.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2018-2021 Splunk Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package updatecr
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/splunk/splunk-operator/test/testenv"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+var _ = Describe("update cr test", func() {
+
+	var deployment *testenv.Deployment
+	var defaultCPULimits string
+	var newCPULimits string
+
+	BeforeEach(func() {
+		var err error
+		deployment, err = testenvInstance.NewDeployment(testenv.RandomDNSName(3))
+		Expect(err).To(Succeed(), "Unable to create deployment")
+		defaultCPULimits = "4"
+		newCPULimits = "2"
+	})
+
+	AfterEach(func() {
+		// When a test spec failed, skip the teardown so we can troubleshoot.
+		if CurrentGinkgoTestDescription().Failed {
+			testenvInstance.SkipTeardown = true
+		}
+		if deployment != nil {
+			deployment.Teardown()
+		}
+	})
+
+	Context("Standalone deployment (S1)", func() {
+		It("update: can deploy a standalone instance, change its CR, update the instance", func() {
+
+			// Deploy Standalone
+			standalone, err := deployment.DeployStandalone(deployment.GetName())
+			Expect(err).To(Succeed(), "Unable to deploy standalone instance")
+
+			// Verify Standalone goes to ready state
+			testenv.StandaloneReady(deployment, deployment.GetName(), standalone, testenvInstance)
+
+			// Verify CPU limits before updating the CR
+			standalonePodName := fmt.Sprintf(testenv.StandalonePod, deployment.GetName(), 0)
+			testenv.VerifyCPULimits(deployment, testenvInstance.GetName(), standalonePodName, defaultCPULimits)
+
+			// Change CPU limits to trigger CR update
+			standalone.Spec.Resources.Limits = corev1.ResourceList{
+				"cpu": resource.MustParse(newCPULimits),
+			}
+			err = deployment.UpdateCR(standalone)
+			Expect(err).To(Succeed(), "Unable to deploy standalone instance with updated CR ")
+
+			// Verify Standalone is updating
+			testenv.VerifyStandaloneUpdating(deployment, deployment.GetName(), standalone, testenvInstance)
```

Comment:
I am guessing this was needed to ensure we wait for the change to kick-in & get applied, before verifying the new CPU limits

Reply:
All the verification methods seem to be **wait till success with a timeout**. Since the eventual goal of the test here is to ensure that the Standalone is Ready and then verify CPU limits I thought we could avoid it. If there is a transient period where K8S takes time to update this is fine I guess.

Reply:
As of now we need that VerifyStandaloneUpdating function because the next function StandaloneReady behaves as it will directly fail as it expects 'Ready' but got 'Updating' instead.